### PR TITLE
bumped to version 2.0

### PR DESCRIPTION
**Pull request by anestos**
- Fixed an issue caused by json encoding of a base64 encoded string, which was not accepted by Freja API
- Unified the template and added support for swedish language
- Added the possibility to login using phoneNumber
- Added configuration option to set the minimum required registration level
- Added configuration option to select which attributes to be returned by Freja API
- Added support for Cancel

### CODE REVIEW

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 1]
```diff
@@ -142,45 +146,72 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
         if (statusCode != 200)
         {
-            if (_logger.isErrorEnabled)
+            if (_logger.isWarnEnabled)
             {
                 _logger.warn("Got error response from authentication endpoint: error = {}, {}", statusCode,
                         httpResponse.body(HttpResponse.asString()))
             }
 
+            if (_json.fromJson(httpResponse.body(HttpResponse.asString()))["code"] == 2000)
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:156]: 2000 is right? It's not a typo and should be 200?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 2]
```diff
@@ -142,45 +146,72 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
     }
-    
+
+    private fun buildJsonAuthRequest(postData: Map<String, Any>): String
+    {
+        return "{" + postData.map {
+            val value = when
+            {
+                it.value is String -> "\"" + it.value + "\""
+                else -> it.value
```
Comment (travisspencer) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt]: here you could use `Json` just in case.

**Review by travisspencer — APPROVED** [review]

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt (hunk 1 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:156]: Yes, 2000 is correct, as described here: https://frejaeid.com/rest-api/Authentication%20Services.html

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 3]
```diff
@@ -47,59 +52,59 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
         {
             val username = _userPreferencesManager.username
 
-            if (username != null) {
-                if (_userInfoType == UserInfoType.SSN) {
-                    dataMap["username"] = username
-                } else {
-                    dataMap["email"] = username
-                }
+            if (username != null)
```
Comment (pmhsfelix) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:72]: Why this change?

Comment (pmhsfelix) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:72]: Is that because now the email also uses the `username` field?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 4]
```diff
@@ -142,45 +147,71 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
     }
-    
+
+    private fun buildJsonAuthRequest(postData: Map<String, Any>): String
+    {
+        return "{" + postData.map {
+            val value = when
+            {
+                it.value is String -> "\"" + it.value + "\""
+                else -> _json.toJson(it.value)
```
Comment (pmhsfelix) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:173]: Assuming this `buildJsonAuthRequest` was used to avoid additional encoding on strings, the same issue will happen if there is a string somewhere inside `it.value` (because the recursion uses `_json .toJson` and not `buildJsonAuthRequest`).

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt [hunk 5]
```diff
@@ -142,45 +147,71 @@ class StartRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig,
...
             dataMap["userInfo"] = Base64.getEncoder().encodeToString(_json.toJson(userInfo).toByteArray())
         }
+        else
+        {
+            dataMap["userInfo"] = username
+        }
+
+        if (_minRegistrationLevel == RegistrationLevel.BASIC && _attributesToReturn.contains(AttributesToReturn.EMAIL_ADDRESS))
+        {
+            dataMap["attributesToReturn"] = Collections.singletonList(mapOf("attribute" to AttributesToReturn.EMAIL_ADDRESS))
```
Comment (pmhsfelix) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt]: This singleton list and the contained map could be static.

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt [hunk 6]
```diff
@@ -85,18 +93,30 @@ class WaitRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig)
             else -> false
         }
 
-        if (moveOn)
+        val cancel = when
```
Comment (pmhsfelix) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt:96]: Or not a simple `if`?

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt [hunk 7]
```diff
@@ -198,9 +290,38 @@ class WaitRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig)
         return Optional.empty()
     }
 
+    private fun cancelAuthentication()
+    {
+        val authRef = _sessionManager.get(SESSION_AUTH_REF)?.let {
+            _json.toJson(Collections.singletonMap("authRef", it.value)).toByteArray()
+        }
+                ?: throw config.exceptionFactory.badRequestException(ErrorCode.INVALID_SERVER_STATE,
```
Comment (pmhsfelix) [inline src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt]: Seems a strange indentation, doesn't it.

**Review by pmhsfelix — APPROVED** [review]
Approved with minor comments. However I'm not familiar with the previous code nor with the freja authenticator.

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt (hunk 3 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:72]: yes i removed the logic that changes the name of the field

File: src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt (hunk 4 above)
Comment (anestos) [reply src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/StartRequestHandler.kt:173]: That is true but it still avoid the problem which originally caused this rework (Freja not accepting urlencoded base64encoded strings. Not sure if it is worth spending time safeguarding this more.
